Update rescale_intensity to prevent under/overflow and produce proper output dtype

jni opened the pull request:
## Description

Fixes #4583 

- All computations are now in float, preventing under/overflow
- If `out_range` is a dtype, then the output will be in that dtype (currently it is the input dtype, which is insane).
- If `out_range` is a range of numbers, the output will be of type float and the caller will be responsible for converting to the correct dtype. **This is a breaking change in some circumstances.** See our use of `rescale_intensity` within `equalize_adapthist` which I had to fix.

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] ~~Gallery example in `./doc/examples` (new features only)~~
- [ ] ~~Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark~~
- [x] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

emmanuelle commented on skimage/exposure/exposure.py:
```diff
@@ -266,6 +266,39 @@ def intensity_range(image, range_values='image', clip_negative=False):
+        except TypeError:  # uint10, uint12, uint14
+            # otherwise, return uint16
+            return np.uint16
+
```
add an else here to raise a `TypeError` for example in the case of typos (`flat` instead of `float`). `output_dtype` should be None if none of the ifs is True, and `np.array([1, 2, 3], dtype=None)` does not error...

lagru replied:
I understand and support the first part of your suggestion. 

But why should `output_dtype` be `None`. The way I understand the new behavior is 
- `out_range` is an actual range -> output dtype is `float64`
- `out_range` is already a dtype (corresponding to the input dtype) -> use same type for output

`None` would just result in the returned output being of dtype `float64` wouldn't it?

emmanuelle replied:
@lagru probably what I wrote was not clear, please tell me if it's clearer now. The function should never return None, it should return a valid value or error out.

lagru replied:
Ah, then we are in agreement. :+1:

emmanuelle commented:
One possibility we did not consider would be to accept for `output_range` a tuple with 3 elements, the last one being a dtype, for example `(0, 200, np.uint8)`. Not sure we want this, but this could avoid an `astype` call when wanting a specific dtype and not floats.

lagru commented on skimage/exposure/exposure.py:
```diff
@@ -266,6 +266,39 @@ def intensity_range(image, range_values='image', clip_negative=False):
     return i_min, i_max
 
 
+def _output_dtype(out_range):
```
I feel like we could choose a more general name that covers `input_dtype` as well. Maybe `dtype_or_range`? Might be a little easier to read.

jni replied:
@lagru 

1. the function name should relate to its return value (always a dtype), not its input parameter! So in this case I think you would want `_dtype_from_dtype_or_range`, which is quite a mouthful!
2. The input dtype is never in question, because it comes from the input array, which is given. So I don't think it's super useful to generalise this for now.

jni replied:
LOL **ignore me**, I just saw that you were referring to the argument. :man_facepalming: I shouldn't comment on PRs before I've finished my morning coffee! :coffee: :sweat_smile:

lagru commented on skimage/exposure/exposure.py:
```diff
@@ -341,10 +374,15 @@ def rescale_intensity(image, in_range='image', out_range='dtype'):
 
     """
-    dtype = image.dtype.type
+    input_dtype = image.dtype.type
```
This is only used once now, maybe move it into the if-clause or remove the assignment altogether?

lagru commented on skimage/exposure/tests/test_exposure.py:
```diff
@@ -237,10 +237,16 @@ def test_rescale_in_range_clip():
     out = exposure.rescale_intensity(image, out_range=(0, 127))
-    assert out.dtype == np.int8
-    assert_array_almost_equal(out, [0, 63, 127])
+    assert out.dtype == np.float_
```
```suggestion
    assert out.dtype == np.float64
```
I find this more explicit or is there a reason you prefer `float_`? Same in the test below...

jni replied:
I prefer `float_` because it is the default float value for that platform. In some platforms, `np.zeros(a_shape)` returns float32, not float64. I think we should return this default type, and that is what `float_` evaluates to.

lagru replied:
Ah that makes sense. :+1:  Strange that my search pointed me to an old NumPy doc version with the statement "[Shorthand for float64](https://docs.scipy.org/doc/numpy-1.13.0/user/basics.types.html#array-types-and-conversions-between-types)" first.

jni commented:
@emmanuelle @lagru Thank you for the reviews! I've addressed all of your comments.

@emmanuelle 

> One possibility we did not consider would be to accept for output_range a tuple with 3 elements, the last one being a dtype, for example `(0, 200, np.uint8)`

I don't like that solution very much, it seems clunky and it will complicate the code quite a lot. I *would* consider adding `dtype=` as an additional keyword argument, but then the desired behaviour would be unclear if e.g. I pass `out_range='uint16', dtype=np.uint32`. I would say that the `dtype=` argument should override all previous options, but that's just one of many possible alternatives.

At any rate, I think we can discuss this separately and address it in a separate PR, since this is already a big improvement on existing behaviour. None of the above proposals would allow existing code to continue working when the output dtype was incorrectly being inferred from the input dtype (as in my fix for `equalize_adapthist`), so I don't think they are critical to get in now.

lagru commented on skimage/exposure/exposure.py:
```diff
@@ -266,6 +266,50 @@ def intensity_range(image, range_values='image', clip_negative=False):
+    """Determine the output dtype for rescale_intensity.
+
+    The dtype is determined according to the following rules:
+    - if ``out_range`` is a dtype, that is the output dtype.
+    - if ``out_range`` is a dtype string, that is the dtype used, unless it
+      is not a NumPy data type (e.g. 'uint12' for 12-bit unsigned integers),
+      in which case the data type that can contain it will be used
+      (e.g. uint16 in this case).
+    - if ``out_range`` is a pair of values, the output data type will be
```
```suggestion
    - if ``dtype_or_range`` is a dtype, that is the output dtype.
    - if ``dtype_or_range`` is a dtype string, that is the dtype used, unless it
      is not a NumPy data type (e.g. 'uint12' for 12-bit unsigned integers),
      in which case the data type that can contain it will be used
      (e.g. uint16 in this case).
    - if ``dtype_or_range`` is a pair of values, the output data type will be
```

pep8speaks commented:
Hello @jni! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`skimage/exposure/exposure.py`](https://github.com/scikit-image/scikit-image/blob/1e80704e3dcf565a36661cb308467f0e771b6b55/skimage/exposure/exposure.py):

> [Line 274:80](https://github.com/scikit-image/scikit-image/blob/1e80704e3dcf565a36661cb308467f0e771b6b55/skimage/exposure/exposure.py#L274): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

jni commented:
Thanks for keeping me honest, @lagru! 😂

lagru commented on skimage/exposure/_adapthist.py:
```diff
@@ -75,7 +75,9 @@ def equalize_adapthist(image, kernel_size=None,
 
     image = img_as_uint(image)
-    image = rescale_intensity(image, out_range=(0, NR_OF_GRAY - 1))
+    image = np.round(
+        rescale_intensity(image, out_range=(0, NR_OF_GRAY - 1))
+    ).astype(np.uint16)
```
Do we need to worry that the rounding here is a breaking change? See the changed unit test below..

jni replied:
@lagru Ah, yes, I'd meant to discuss that. imho rounding is the correct behaviour and I would consider this a bug fix, but we should definitely discuss it! @scikit-image/core any thoughts on this?

rfezzani replied:
I agree with @jni that the the rounding is the correct behavior. The best that we can do is may be to add a note in the documentation about this breaking change.

alexdesiqueira replied:
> imho rounding is the correct behaviour and I would consider this a bug fix

+1.

jni replied:
I've added a note to the release notes and a versionchanged marker to both functions.

m-albert commented:
Thanks for the fix! After this is merged I'd raise another issue (I think independent of this one) on this function with a proposed fix. Cheers

jni commented:
@m-albert why do you need to wait until this is merged?

m-albert commented:
@jni Somehow I was assuming it would be cleaner to address it once this is merged, but actually I can just open the issue, sorry for the mystery. See https://github.com/scikit-image/scikit-image/issues/4596

jni commented:
@scikit-image/core I've addressed the concerns about changed behaviours by adding more explicit notes in the release notes, as well as `.. versionchanged::` markers in the docstrings. Pending CI I think this is ready to be merged!

emmanuelle commented:
Thank you @jni, merging!
